### PR TITLE
Storage: row ID generation for MVCC bitmap filter

### PR DESCRIPTION
**Pull request by JinheLin**
### What problem does this PR solve?

Issue Number: ref #6296 

Problem Summary:

- In order to build a MVCC bitmap filter, we need to scan three columns of handle, version and tag.
- Iterating through the input stream of `Segment::getInputStream`  can obtain an ordered data stream.
- However, the current input stream does not contain the position of the record on the array, so a "virtual column" needs to be added to indicate the record position. This "virtual column" is called `SegmentRowIdCol`.
- In order to generate the `SegmentRowIdCol`, the read interfaces of Delta and Stable need to return the 'row position' of returned data.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/DeltaMerge/DeltaMerge.h
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/DeltaMerge/DeltaMerge.h:299]: how about moving the two function to inside `class Block`?

File: dbms/src/Storages/DeltaMerge/DeltaMerge.h
Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/DeltaMerge.h:95]: get, thanks

File: dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileSetReader.h
Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileSetReader.h:58]: why need default paremeter? just pass `{}` when `row_id` is not needed?

File: dbms/src/Storages/DeltaMerge/Delta/Snapshot.cpp
Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/Delta/Snapshot.cpp:123]: From my previous test, for `vector<bool>`, using the algorithm from `std::algorithm` is much faster than the loop. Not quite sure about `UInt32`

File: dbms/src/Storages/DeltaMerge/DeltaMerge.h
Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/DeltaMerge.h]: Since we need fillSegmentRowId before the filters, move inside `class Block` can help reuse the logic.

Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/DeltaMerge.h]: When implementing late materialization, we need SegmentRowId to generate fitering-bitmap.

File: dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileSetReader.cpp
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileSetReader.cpp:170]: usually, when MVCC, the remaining rows are more than the filter-out rows, maybe we can record the filter-out row_id?

File: dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileSetReader.h
Comment (flowbehappy) [reply dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileSetReader.h:58]: Pointer is good for me. `std::optional<std::vector<UInt32>> &` is more complicated.

Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileSetReader.h:58]: make sense, thank you~

File: dbms/src/Storages/DeltaMerge/RowKeyFilter.h
Comment (flowbehappy) [inline dbms/src/Storages/DeltaMerge/RowKeyFilter.h:180]: It looks like a bug here because the result of `filter` is not used.

File: dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileBig.cpp
Comment (flowbehappy) [reply dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileBig.cpp:142]: `p0 p1, p2, p3`
There are possibilities that `range` can filter out `p0` and part of `p1`. Although it won't happen by now, I am afraid it could be in the future after somebody changes the code.
I mean we better don't rely on the assumption that `range` can only filter out part of `p0` or `p3`, it is dangerous.

Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileBig.cpp:142]: like
`p0: [0, 1000), p1:[1000, 2000), p2:[2000, 3000), p3:[3000, 4000)`
`range: [0, 1003)`
?

File: dbms/src/Storages/DeltaMerge/BitmapFilter/BitmapFilterBlockInputStream.cpp
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/DeltaMerge/BitmapFilter/BitmapFilterBlockInputStream.cpp:60]: optimize here, when we call `bitmap_filter->get` also return the positive count as `positive`, when `positive == block.rows()` skip the filter, and call `col.column = col.column->filter(filter, positive);` instead

File: dbms/src/Interpreters/Settings.h
Comment (flowbehappy) [inline dbms/src/Interpreters/Settings.h:323]: We should remove this setting before merging this pr.

File: dbms/src/Storages/DeltaMerge/BitmapFilter/BitmapFilter.cpp
Comment (flowbehappy) [inline dbms/src/Storages/DeltaMerge/BitmapFilter/BitmapFilter.cpp:41]: `f == nullptr` means all positive seems not human-friendly to me.
Besides, better use `!f` instead of `f == nullptr`.

Comment (flowbehappy) [reply dbms/src/Storages/DeltaMerge/BitmapFilter/BitmapFilter.cpp:41]: You can use another function like `setAll` to make all rows positive. And do this logic outside of this class.

File: dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileSetReader.h
Comment (flowbehappy) [inline dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileSetReader.h:58]: Please add documents about `row_ids`.

File: dbms/src/Storages/DeltaMerge/DeltaMerge.h
Comment (flowbehappy) [inline dbms/src/Storages/DeltaMerge/DeltaMerge.h:555]: I wonder why delta's `row_ids` cannot be calculated like stable's.

File: dbms/src/Storages/DeltaMerge/Segment.cpp
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/DeltaMerge/Segment.cpp:2324]: I think `enable_del_clean_read` is always false. TiDB Planner is not aware of the tag column, so columns_to_read never contain tag column.

**Review by Lloyd-Pottiger — COMMENTED** [review]
The code is mixed with bitmap logic now.

File: dbms/src/Storages/DeltaMerge/BitmapFilter/BitmapFilterBlockInputStream.cpp
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/DeltaMerge/BitmapFilter/BitmapFilterBlockInputStream.cpp:63]: what if `all_match = true && return_filter = true`

File: dbms/src/Storages/DeltaMerge/DeltaMerge.h
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/DeltaMerge/DeltaMerge.h:297]:
```suggestion
        for (UInt32 i = 0; i < row_ids.size(); ++i)
```

File: dbms/src/Storages/DeltaMerge/BitmapFilter/BitmapFilter.cpp
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/DeltaMerge/BitmapFilter/BitmapFilter.cpp:44]: Feel confused about this function. When the stream is `DMVersionFilterBlockInputStream`, the return FilterPtr is associated with the block return by `DMRowKeyFilterBlockInputStream` rather than the original block in dmfile, why it can be used for generating bitmap?

File: dbms/src/Storages/DeltaMerge/BitmapFilter/BitmapFilter.h
Comment (flowbehappy) [reply dbms/src/Storages/DeltaMerge/BitmapFilter/BitmapFilter.h:46]: `BitmapFilter` better be a pure data object. It is not necessary to make `snap` a member of it. Besides, `snap` is only here to be returned by `snapshot`, it does not mean anything to this object. 

I think you can consider passing `snap` around outside of this object.

File: dbms/src/Storages/DeltaMerge/DMVersionFilterBlockInputStream.cpp
Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/DMVersionFilterBlockInputStream.cpp:415]: ok

File: dbms/src/Storages/DeltaMerge/Segment.cpp
Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/Segment.cpp:2436]: We need to fill `segment_row_id_col` with continuous row_id, and the filter twice. We can call `Block read(FilterPtr, true)` both in `DMRowKeyFilterBlockInputStream` and `DMVersionFilterBlockInputStream`, which means that there is no need to fill `segment_row_id_col` with continuous row_id, and filter only once.

Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/Segment.cpp:2436]: Since there is no `DeltaMergeBlockInputStream`, the order of rows will not change, so we can do this optimization.

File: dbms/src/Storages/DeltaMerge/BitmapFilter/BitmapFilter.cpp
Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/BitmapFilter/BitmapFilter.cpp:44]: get it

File: dbms/src/Storages/DeltaMerge/BitmapFilter/BitmapFilter.cpp
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/DeltaMerge/BitmapFilter/BitmapFilter.cpp:111]: if `all_match`, we can just clear `filter`?

File: dbms/src/Storages/DeltaMerge/BitmapFilter/BitmapFilter.cpp
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/DeltaMerge/BitmapFilter/BitmapFilter.cpp:124]: The return string usually is too long, I think we can just return size and number of positive.

Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/BitmapFilter/BitmapFilter.cpp:124]: ok, get

File: dbms/src/Storages/DeltaMerge/Segment.cpp
Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/Segment.cpp:2436]: Are there some special reason for not supporting `read(FilterPtr, true)` in `DMRowKeyFilterBlockInputStream`? Maybe   we can achieve it since filter is heavy-cost?

File: dbms/src/Storages/DeltaMerge/BitmapFilter/BitmapFilter.cpp
Comment (flowbehappy) [inline dbms/src/Storages/DeltaMerge/BitmapFilter/BitmapFilter.cpp:59]:
```suggestion
```

File: dbms/src/Storages/DeltaMerge/BitmapFilter/BitmapFilter.cpp
Comment (flowbehappy) [inline dbms/src/Storages/DeltaMerge/BitmapFilter/BitmapFilter.cpp:93]:
```suggestion
```

File: dbms/src/Storages/DeltaMerge/BitmapFilter/BitmapFilter.h
Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/BitmapFilter/BitmapFilter.h:46]: Agree, we can move `BitmapFilter` to `libs/libcommon` or `dbms/src/Common`, and set `avx2` flag during compiling to speed up.

File: dbms/src/Storages/DeltaMerge/Segment.cpp
Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/Segment.cpp:2436]: I think it is worth to do this now. We do not need to `createSegmentRowIdCol` in `ConcatSkippableBlockInputStream`. When the parent of `ConcatSkippableBlockInputStream` is `DeltaMergeBlockInputStream`, there is no need to `createSegmentRowIdCol` since  `DeltaMergeBlockInputStream` will do that. And the parent of `ConcatSkippableBlockInputStream` is `DMRowKeyFilterBlockInputStream`, we can call `read(FilterPtr, true)` to return filter, and we can use filter to generaate bitmap.

**Review by flowbehappy — APPROVED** [review]
LGTM

File: dbms/src/Storages/DeltaMerge/Segment.cpp
Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/Segment.cpp:2436]: no, I mean use filter to replace `SegmentRowIdCol`.

`ConcatSkippableBlockInputStream` return the following block:

```
startOffset: 100
col filter
1   1
1   1
2   1
3   1
4   1
5   1
6   1
```

After `DMVersionFilterBlockInputStream`, the block is:

```
startOffset: 100
col filter
1   0
1   1
2   1
3   1
4   1
5   1
6   1
```

After `DMRowKeyFilterBlockInputStream`, the block is:

```
startOffset: 100
col filter
1   0
1   0
2   1
3   1
4   1
5   1
6   0
```

Then we can just set [102, 103, 104, 105] of bitmap to 1.

Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/Segment.cpp:2436]: > However, read continuous packs is not necessary since the packs to be read has been specified. We can read multiple packs from DMFile concurrently in this situation in the future or in the cloud native mode, we can return several already cached packs first.

But fill `SegmentRowIdCol` in `ConcatSkippableBlockInputStream` can not handle this case either. Like we can a block in `ConcatSkippableBlockInputStream`, which contain data in `pack2`, `pack9` and `pack88`, how we can know about that?

Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/Segment.cpp:2436]: > because it would introduce special code paths and more restrictions but the benefits are small.

fill `SegmentRowIdCol` in `ConcatSkippableBlockInputStream` also introduce code path 🤔 And using `offset` only need little changes in `BitmapFilter`

**Review by Lloyd-Pottiger — APPROVED** [review]
The others LGTM. We can add some comments about previous discussion. Leave it to be optimized.

**Comment by Lloyd-Pottiger** [comment]
> > The others LGTM. We can add some comments about previous discussion. Leave it to be optimized.
> 
> Can you write an issue about it? @Lloyd-Pottiger

ok

**Pull request merged**
